### PR TITLE
Add cadical SAT solver to cmake build

### PR DESCRIPTION
**Pull request by mgudemann**
Main changes

- remove use of named varargs in macro, replace `...(FOO,ARGS...)` with `...(__VA_ARGS__)` where possible due to comma bug in C++ varags macros
- remove `;` from closing bracket in namespace definitons and `extern C { }` declarations
- add fixed `config.hpp` (creating git id using CBMC's makes no sense)

### CODE REVIEW

**Comment by mgudemann** [comment]
@tautschnig please have a look if cadical still works as expected in `make` builds, my tests may be more limited than your use

@others
I am specifically interested in getting hints of how things can be done better in CMake and how to best simplify the patch file

File: scripts/cadical_CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    src/restart.cpp
+    src/signal.cpp
+    src/solution.cpp
+    src/stats.cpp
+    src/subsume.cpp
+    src/transred.cpp
+    src/util.cpp
+    src/var.cpp
+    src/vivify.cpp
+    src/watch.cpp
```
Comment (tautschnig) [inline scripts/cadical_CMakeLists.txt:49]: Having to explicitly list the files of an external project seems rather flaky?

File: scripts/cadical-patch [hunk 2]
```diff
@@ -1,10 +1,1140 @@
...
+--- /dev/null
++++ b/src/config.hpp
+@@ -0,0 +1,7 @@
++#define CADICAL_VERSION "06w"
++#define CADICAL_GITID "N/A"
++#define CADICAL_CXX "N/A"
++#define CADICAL_CXXFLAGS "N/A"
++#define CADICAL_CXXVERSION "N/A"
++#define CADICAL_COMPILED "N/A"
++#define CADICAL_OS "N/A"
```
Comment (tautschnig) [inline scripts/cadical-patch:212]: Why is this needed? Can't you run `configure`?

**Review by tautschnig — COMMENTED** [review]
I would suggest to flag all the issues in the upstream project (you might want to add more requests to https://github.com/arminbiere/cadical/issues/4). Do keep in mind that CaDiCaL at present can only be used with --stop-on-fail - incremental use is not yet supported.

**Comment by mgudemann** [comment]
Michael Tautschnig <notifications@github.com> writes:

> tautschnig commented on this pull request.
>
> I would suggest to flag all the issues in the upstream project (you might want to add more requests to https://github.com/arminbiere/cadical/issues/4). Do keep in mind that CaDiCaL at present can only be used with --stop-on-fail - incremental use is not yet supported.

ok, I will the closing `};` issues. The named vararg macros are GNU
extensions, so ok to use and problematic just for our use.

Snap, I missed that it doesn't support incremental solving. That voids
the intended use for now.

> +@@ -0,0 +1,7 @@
> ++#define CADICAL_VERSION "06w"
> ++#define CADICAL_GITID "N/A"
> ++#define CADICAL_CXX "N/A"
> ++#define CADICAL_CXXFLAGS "N/A"
> ++#define CADICAL_CXXVERSION "N/A"
> ++#define CADICAL_COMPILED "N/A"
> ++#define CADICAL_OS "N/A"
>
> Why is this needed? Can't you run `configure`?

Yes, but the actual CXX, CXX_VERSION and CXXFLAGS are not those from the
configure step, but those used by cmake in compilation. Also, we are not
in cadical's git, so it would insert CBMC git shasum as id.

On the other hand fixing config.hpp means we have to run one less
external program from cmake.

**Comment by tautschnig** [comment]
> ok, I will the closing `};` issues. The named vararg macros are GNU extensions, so ok to use and problematic just for our use.

Do note https://github.com/arminbiere/cadical/blob/master/CONTRIBUTING - you may have been too kind...

> Snap, I missed that it doesn't support incremental solving. That voids the intended use for now.

Hopefully Armin will fix this fairly soon - it's documented at https://github.com/arminbiere/cadical/issues/2.

**Comment by mgudemann** [comment]
> ok, I will the closing }; issues. The named vararg macros are GNU extensions, so ok to use and problematic just for our use.
>
> Do note https://github.com/arminbiere/cadical/blob/master/CONTRIBUTING - you may have been too kind...

surprising, but well, why not

>
> Snap, I missed that it doesn't support incremental solving. That voids the intended use for now.
>
> Hopefully Armin will fix this fairly soon - it's documented at arminbiere/cadical#2.

would be nice to see how it fares with TG / string-solver once that's done

**Comment by mgudemann** [comment]
closing for now as incremental solving is required
